File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */

// Method ... generate all the path sum of the tree and then return the maxx
// path of the given sum...

class Solution {
public:
    int maxSum;
    int solve(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }
        int left = solve(root->left);
        int right = solve(root->right);

        int neeche_hi_mil_gaya_answer = left + right + root->val; // (1)
        int koi_ek_accha = max(left, right) + root->val;          // (2)
        int only_root_accha = root->val;                          // (3)
        maxSum =max({maxSum,neeche_hi_mil_gaya_answer, koi_ek_accha, only_root_accha});
        return max(koi_ek_accha,only_root_accha);
    }
    int maxPathSum(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }
        maxSum = INT_MIN;
        solve(root);
        return maxSum;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        maxSum = float("-inf") # make sure to strat with the negative infinity ..
        def Solve(node):
            nonlocal maxSum 
            if node is None:
                return 0
            left = Solve(node.left)
            right = Solve(node.right)

            summ1 = left + right + node.val #case - 1
            summ2 = max(left, right) + node.val # case - 2
            summ3 = node.val # case - 3
            maxSum = max(maxSum, summ1, summ2, summ3)

            return max(summ2, summ3)

        Solve(root)

        return maxSum
